src/libraries/Directory.Build.targets

```diff
@@ -218,8 +218,9 @@
     </When>
   </Choose>
 
+  <!-- Enable the SkipLocalsInit feature for any source project that is part of the .NETCoreApp shared framework. -->
   <PropertyGroup>
-    <SkipLocalsInit Condition="'$(SkipLocalsInit)' == '' and '$(MSBuildProjectExtension)' == '.csproj' and '$(IsNETCoreAppSrc)' == 'true' and ($([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', '$(NetCoreAppCurrent)')))">true</SkipLocalsInit>
+    <SkipLocalsInit Condition="'$(SkipLocalsInit)' == '' and '$(MSBuildProjectExtension)' == '.csproj' and '$(IsNETCoreAppSrc)' == 'true' and $([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net5.0'))">true</SkipLocalsInit>
```

Comment:
This still makes .NET 6 assemblies SkipLocalsInit?

What starts being included that shouldn't if we remove the entire last condition?

Reply:
The SkiplocalsInit optimization was added for libraries during NET5, which is why I didn't apply it to netcoreapp3.1 assets.

> This still makes .NET 6 assemblies SkipLocalsInit?

Yes this applies to net5.0 upwards.

Reply:
> The SkiplocalsInit optimization was added for libraries during NET5, which is why I didn't apply it to netcoreapp3.1 assets.

We used the SkipLocalsInit attribute for it starting in .NET 5, but we were using the linker's support for stripping out locals init well before that, e.g.
https://github.com/dotnet/corefx/blob/18acb8f3709c194a2debd6be8abe274e860869f6/illink.targets#L109-L114

Reply:
Then we likely want this to apply to netcoreapp3.1 assets as well as the linker doesn't do that anymore, not even for netcorapp3.1 assets.

Reply:
> Then we likely want this to apply to netcoreapp3.1 assets as well as the linker doesn't do that anymore, not even for netcorapp3.1 assets.

[SkipLocalsInit] was added in .NET 5.  We'd presumably need to inject a private copy of the attribute to make that happen.  Do we already do that in some cases?

Reply:
Updated.

Reply:
> [SkipLocalsInit] was added in .NET 5. We'd presumably need to inject a private copy of the attribute to make that happen. Do we already do that in some cases?

The allconfigurations build now warns about what you pointed out earlier. So apparently yes, we need to pass  `SkipLocalsInitAttribute` as a compile source for netcoreapp3.1 targets. We do similar for nullable attributes, obsoletion apis and platform attributes: https://github.com/dotnet/runtime/blob/84ca939dc9931e8b4d5ab4197cdc99a4621984a6/src/libraries/Directory.Build.targets#L190-L219.